File: ZezeJava/ZezexJava/server/src/Game/Bag/ModuleBag.java
package Game.Bag;

import Zeze.Arch.ProviderUserSession;
import Zeze.Transaction.*;
import Game.*;
import Zeze.Transaction.Collections.LogMap2;

//ZEZE_FILE_CHUNK {{{ IMPORT GEN
//ZEZE_FILE_CHUNK }}} IMPORT GEN

public final class ModuleBag extends AbstractModule {
	public void Start(App app) {
		_tbag.getChangeListenerMap().AddListener(new BagChangeListener());
	}

	public void Stop(App app) {
	}

	private static class BagChangeListener implements ChangeListener {
		private static final String Name = "Game.Bag";
		public static String getName() {
			return Name;
		}

		@Override
		public final void OnChanged(Object key, Changes.Record c) {
			switch (c.getState()) {
			case Changes.Record.Put:
				// 记录改变，通知全部。
				BBag bbag = (BBag)c.getValue();
				var sbag = new SBag();
				Bag.ToProtocol(bbag, sbag.Argument);
				Game.App.getInstance().getProvider().Online.sendReliableNotify((Long)key, getName(), sbag);
				break;

			case Changes.Record.Edit:
				// 增量变化，通知变更。
				@SuppressWarnings("unchecked")
				var notemap2 = (LogMap2<Integer, BItem>)c.getVariableLog(tbag.VAR_Items);
				if (null != notemap2) {
					notemap2.MergeChangedToReplaced();
					SChanged changed = new SChanged();
					changed.Argument.setChangeTag(BChangedResult.ChangeTagNormalChanged);
					changed.Argument.getItemsReplace().putAll(notemap2.getReplaced());
					for (var p : notemap2.getRemoved()) {
						changed.Argument.getItemsRemove().add(p);
					}
					Game.App.getInstance().getProvider().Online.sendReliableNotify((Long)key, getName(), changed);
				}
				break;

			case Changes.Record.Remove:
				SChanged changed2 = new SChanged();
				changed2.Argument.setChangeTag(BChangedResult.ChangeTagRecordIsRemoved);
				Game.App.getInstance().getProvider().Online.sendReliableNotify((Long)key, getName(), changed2);
				break;
			}
		}
	}

	// protocol handles
	@Override
	protected long ProcessMoveRequest(Move rpc) throws Throwable {
		var session = ProviderUserSession.Get(rpc);
		// throw exception if not login
		var moduleCode = GetBag(session.getRoleId().longValue()).Move(
				rpc.Argument.getPositionFrom(), rpc.Argument.getPositionTo(), rpc.Argument.getNumber());
		if (moduleCode != 0) {
			return ErrorCode(moduleCode);
		}
		session.SendResponse(rpc);
		return 0;
	}

	@Override
	protected long ProcessDestroyRequest(Destroy rpc) throws Throwable {
		var session = ProviderUserSession.Get(rpc);
		var moduleCode = GetBag(session.getRoleId().longValue()).Destory(rpc.Argument.getPosition());
		if (0 != moduleCode) {
			return ErrorCode(moduleCode);
		}
		session.SendResponse(rpc);
		return 0;
	}

	@Override
	protected long ProcessSortRequest(Sort rpc) throws Throwable {
		var session = ProviderUserSession.Get(rpc);
		Bag bag = GetBag(session.getRoleId().longValue());
		bag.Sort(null);
		session.SendResponse(rpc);
		return Procedure.Success;
	}

	@Override
	protected long ProcessGetBagRequest(GetBag rpc) throws Throwable {
		var session = ProviderUserSession.Get(rpc);

		GetBag(session.getRoleId().longValue()).ToProtocol(rpc.Result);
		session.SendResponse(rpc);
		Game.App.getInstance().getProvider().Online.addReliableNotifyMark(session.getRoleId().longValue(), BagChangeListener.getName());
		return Procedure.Success;
	}

	// for other module
	public Bag GetBag(long roleid) {
		return new Bag(roleid, _tbag.getOrAdd(roleid));
	}

	@Override
	protected long ProcessCUse(CUse protocol) throws Throwable {
		var session = ProviderUserSession.Get(protocol);
		Bag bag = GetBag(session.getRoleId().longValue());
		var item = bag.GetItem(protocol.Argument.getPosition());
		if (null != item && item.Use()) {
			if (bag.Remove(protocol.Argument.getPosition(), item.getId(), 1)) {
				return Procedure.Success;
			}
		}
		return Procedure.LogicError;

	}

	// ZEZE_FILE_CHUNK {{{ GEN MODULE
    public ModuleBag(Game.App app) {
        super(app);
    }
	// ZEZE_FILE_CHUNK }}} GEN MODULE
}

File: Sample/server/Game/Bag/ModuleBag.cs

using System.Threading.Tasks;
using Zeze.Arch;
using Zeze.Net;
using Zeze.Transaction;
using Zeze.Transaction.Collections;
using Zeze.Util;

namespace Game.Bag
{
    public sealed partial class ModuleBag : AbstractModule
    {
        public void Start(Game.App app)
        {
            _tBag.ChangeListenerMap.AddListener(new BagChangeListener());
        }

        public void Stop(Game.App app)
        {
        }

        class BagChangeListener : ChangeListener
        {
            public static string Name { get; } = "Game.Bag";

            public void OnChanged(object key, Bean value)
            {
                // 记录改变，通知全部。
            }

            public void OnChanged(object key, Changes.Record changes)
            {
                switch (changes.State)
                {
                    case Changes.Record.Remove:
                        {
                            var changed = new SChanged();
                            changed.Argument.ChangeTag = BChangedResult.ChangeTagRecordIsRemoved;
                            Game.App.Instance.ProviderImplementWithOnline.Online.SendReliableNotify((long)key, Name, changed);
                        }
                        break;

                    case Changes.Record.Put:
                        {
                            var bbag = (BBag)changes.Value;
                            var sbag = new SBag();
                            Bag.ToProtocol(bbag, sbag.Argument);
                            Game.App.Instance.ProviderImplementWithOnline.Online.SendReliableNotify((long)key, Name, sbag);
                        }
                        break;

                    case Changes.Record.Edit:
                        var logbean = changes.GetLogBean();
                        if (logbean.Variables.TryGetValue(tBag.VAR_Items, out var log))
                        {
                            var note = (LogMap2<int, BItem>)log;
                            note.MergeChangedToReplaced();

                            var changed = new SChanged();
                            changed.Argument.ChangeTag = BChangedResult.ChangeTagNormalChanged;
                            changed.Argument.ItemsReplace.AddRange(note.Replaced);
                            foreach (var p in note.Removed)
                                changed.Argument.ItemsRemove.Add(p);

                            Game.App.Instance.ProviderImplementWithOnline.Online.SendReliableNotify((long)key, Name, changed);
                        }
                        break;
                }
            }
        }

        // protocol handles
        protected override async Task<long> ProcessMoveRequest(Protocol p)
        {
            var rpc = p as Move;
            var session = ProviderUserSession.Get(rpc);
            // throw exception if not login
            var moduleCode = (await GetBag(session.RoleId.Value)).Move(
                rpc.Argument.PositionFrom,
                rpc.Argument.PositionTo,
                rpc.Argument.Number);
            if (moduleCode != 0)
                return ErrorCode((ushort)moduleCode);
            session.SendResponseWhileCommit(rpc);
            return 0;
        }

        protected override async Task<long> ProcessDestroyRequest(Protocol p)
        {
            var rpc = p as Destroy;
            var session = ProviderUserSession.Get(rpc);
            var moduleCode = (await GetBag(session.RoleId.Value)).Destroy(rpc.Argument.Position);
            if (0 != moduleCode)
                return ErrorCode((ushort)moduleCode);
            session.SendResponseWhileCommit(rpc);
            return 0;
        }

        protected override async Task<long> ProcessSortRequest(Protocol p)
        {
            var rpc = p as Sort;
            var session = ProviderUserSession.Get(rpc);
            Bag bag = await GetBag(session.RoleId.Value);
            bag.Sort();
            session.SendResponseWhileCommit(rpc);
            return ResultCode.Success;
        }

        protected override async Task<long> ProcessGetBagRequest(Protocol p)
        {
            var rpc = p as GetBag;
            var session = ProviderUserSession.Get(rpc);

            (await GetBag(session.RoleId.Value)).ToProtocol(rpc.Result);
            session.SendResponseWhileCommit(rpc);
            await Game.App.Instance.ProviderImplementWithOnline.Online.AddReliableNotifyMark(
                session.RoleId.Value, BagChangeListener.Name);
            return ResultCode.Success;
        }

        // for other module
        public async Task<Bag> GetBag(long roleid)
        {
            return new Bag(roleid, await _tBag.GetOrAddAsync(roleid));
        }

        protected override async Task<long> ProcessCUse(Protocol p)
        {
            var protocol = p as CUse;
            var session = ProviderUserSession.Get(protocol);
            Bag bag = await GetBag(session.RoleId.Value);
            Item.Item item = bag.GetItem(protocol.Argument.Position);
            if (null != item && item.Use())
            {
                if (bag.Remove(protocol.Argument.Position, item.Id, 1))
                    return ResultCode.Success;
            }
            return ResultCode.LogicError;

        }
    }
}
